### PR TITLE
Add emoji picker to menubar (Fixes: #987)

### PR DESCRIPTION
**Pull request by mejo-**
* Resolves: #987
* Target version: master 

### Summary

This is a first attempt to add the emoji picker to the text editor menubar. It works for me, but still has some rough edges.

Here's a screencast: recording

I'm curious to hear your thoughts @juliushaertl and @azul :blush:

The way menu actions are organized in `src/mixins/menubar.js` at the moment unfortunately is a bit limited. Since the emoji picker isn't a simple `<button>` or `<ActionButton>` element and I prefer to reuse the `<EmojiPicker>` component from nextcloud-vue, I went a different route: I added the emoji action in `src/mixins/menubar.js` but special-treated it in the `MenuBar` component. I'm not particularly happy with this approach though. Maybe you have a better idea for a clean implementation?

### CODE REVIEW

File: src/components/MenuBar.vue [hunk 1]
```diff
@@ -127,6 +138,7 @@ export default {
 			forceRecompute: 0,
 			submenuVisibility: {},
 			lastImagePath: null,
+			showEmojiPicker: true,
```
Comment (juliushaertl) [inline src/components/MenuBar.vue:141]: Seems unused

File: src/nodes/Emoji.js [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+import { Node } from 'tiptap'
+import { insertText } from 'tiptap-commands'
+
+export default class Emoji extends Node {
```
Comment (juliushaertl) [inline src/nodes/Emoji.js:26]: I think this should also work with using an extension which would be a more cleaner way from my perspective to handle only the command, see https://github.com/ueberdosis/tiptap/blob/v1/packages/tiptap-extensions/src/extensions/History.js#L41-L48 for example. Unless there is a reason for keeping a dedicated node type of course that I might not see ;)

**Comment by juliushaertl** [comment]
> I'm not particularly happy with this approach though. Maybe you have a better idea for a clean implementation?

Yes, that one is indeed a bit tricky. The reason I went for the array instead of having the buttons directly in the component is that this allows to do some straight forward subset for rendering the popovermenu if the space exceeds the width. I might need to think about that a bit more.

**Comment by szaimen** [comment]
And I think a rebase is probably needed...

File: src/components/MenuBar.vue (hunk 1 above)
Comment (mejo-) [reply src/components/MenuBar.vue:141]: Good catch!

File: src/nodes/Emoji.js (hunk 2 above)
Comment (mejo-) [reply src/nodes/Emoji.js]: You're right, an extension is enough here - no need to make it a node. Changed now :blush:

**Comment by mejo-** [comment]
> Yes, that one is indeed a bit tricky. The reason I went for the array instead of having the buttons directly in the component is that this allows to do some straight forward subset for rendering the popovermenu if the space exceeds the width. I might need to think about that a bit more.

After thinking about it a second time, I'm more confident with my solution of special-treating the EmojiPicker in the for-loop.

Beware though that the way I implemented it results in the EmojiPicker being always displayed as the last non-hidden tool. That's on purpose, given that adding an emoji would require three clicks otherwise (open popovermenu, open emojipicker, select emoji).

Let me know what you think about it :)

**Comment by juliushaertl** [comment]
> > I'm not particularly happy with this approach though. Maybe you have a better idea for a clean implementation?
> 
> Yes, that one is indeed a bit tricky. The reason I went for the array instead of having the buttons directly in the component is that this allows to do some straight forward subset for rendering the popovermenu if the space exceeds the width. I might need to think about that a bit more.

We could probably have a dedicated component with a slot that takes all the actions and then does conditional rendering based on the slot elements but not fully sure if that would be successful.

Something similar how it is done in the actions component https://github.com/nextcloud/nextcloud-vue/blob/master/src/components/Actions/Actions.vue#L649

**Comment by juliushaertl** [comment]
Anyways I'd be fine with the current workaround for now. Doesn't seem too bad ;)

**Comment by mejo-** [comment]
> We could probably have a dedicated component with a slot that takes all the actions and then does conditional rendering based on the slot elements but not fully sure if that would be successful.

To be honest, I'm not sure whether it's worth the effort at the moment :wink:. But you decide. If you'd like it to be that way, I'd give it a try.

> Anyways I'd be fine with the current workaround for now. Doesn't seem too bad ;)

Yay :blush: Feel free to approve/merge. I just changed the PR from draft to ready to merge.

**Comment by mejo-** [comment]
Rebased on latest master.

**Review by juliushaertl — APPROVED** [review]

**Comment by mejo-** [comment]
:heart_eyes:

**Comment by juliushaertl** [comment]
Awesome contribution @mejo- :clap:

**Pull request merged**